core/txpool/bundlepool/bundlepool.go

```diff
@@ -0,0 +1,439 @@
+
+	bundles map[common.Hash]*types.Bundle
+	mu      sync.RWMutex
+
+	slots uint64 // Number of slots currently allocated
+
+	bundleGasPricer *BundleGasPricer
+	simulator       BundleSimulator
+
+	wg         sync.WaitGroup
+	shutdownCh chan struct{}
+}
+
+func New(config Config, chain BlockChain) *BundlePool {
+	// Sanitize the input to ensure no vulnerable gas prices are set
+	config = (&config).sanitize()
+
+	pool := &BundlePool{
+		config:          config,
+		bundles:         make(map[common.Hash]*types.Bundle),
+		bundleGasPricer: NewBundleGasPricer(config.BundleGasPricerExpireTime),
+		shutdownCh:      make(chan struct{}),
+	}
+
+	return pool
+}
+
+func (p *BundlePool) SetBundleSimulator(simulator BundleSimulator) {
+	p.simulator = simulator
+}
+
+func (p *BundlePool) Init(gasTip *big.Int, head *types.Header, reserve txpool.AddressReserver) error {
+	// Set the basic pool parameters
+	p.reset(nil, head)
+	p.SetGasTip(gasTip)
+
+	// Since the user might have modified their pool's capacity, evict anything
+	// above the current allowance
+	for p.slots > p.config.GlobalSlots {
+		p.drop()
+	}
+
+	bundleGauge.Update(int64(len(p.bundles)))
+	slotsGauge.Update(int64(p.slots))
+
+	p.wg.Add(1)
+	go p.loop()
+
+	return nil
+}
+
+// loop is the transaction pool's main event loop, waiting for and reacting to
+// outside blockchain events as well as for various reporting and transaction
+// eviction events.
+func (p *BundlePool) loop() {
+	defer p.wg.Done()
+
+	for {
+		select {
+		case <-p.shutdownCh:
+			return
+		}
+	}
+}
+
+func (p *BundlePool) FilterBundle(bundle *types.Bundle) bool {
+	for _, tx := range bundle.Txs {
+		if !p.filter(tx) {
+			return false
+		}
+	}
+	return true
+}
+
+// AddBundle adds a mev bundle to the pool
+func (p *BundlePool) AddBundle(bundle *types.Bundle) error {
+	if p.simulator == nil {
+		return txpool.ErrSimulatorMissing
+	}
+
+	price, err := p.simulator.SimulateBundle(bundle)
+	if err != nil {
+		return err
+	}
+	minimalGasPrice := p.bundleGasPricer.MinimalBundleGasPrice()
+	if price.Cmp(minimalGasPrice) < 0 {
+		return txpool.ErrBundleGasPriceLow
+	}
+	bundle.Price = price
+
+	hash := bundle.Hash()
+	if _, ok := p.bundles[hash]; ok {
+		return txpool.ErrBundleAlreadyExist
+	}
+	for p.slots+numSlots(bundle) > p.config.GlobalSlots {
+		p.drop()
+	}
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	p.bundles[hash] = bundle
+	p.slots += numSlots(bundle)
+
+	bundleGauge.Update(int64(len(p.bundles)))
+	slotsGauge.Update(int64(p.slots))
+	return nil
+}
+
+func (p *BundlePool) GetBundle(hash common.Hash) *types.Bundle {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	return p.bundles[hash]
+}
+
+func (p *BundlePool) PruneBundle(hash common.Hash) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	p.slots -= numSlots(p.bundles[hash])
+	delete(p.bundles, hash)
+}
+
+func (p *BundlePool) PendingBundles(blockNumber *big.Int, blockTimestamp uint64) []*types.Bundle {
+	ret := make([]*types.Bundle, 0)
+
+	for hash, bundle := range p.bundles {
```

Comment:
Do you need a lock when you iterate `p.bundles`?

Reply:
Removed all the internal call to `PruneBundle`